src/core/lib/channel/promise_based_filter.h

```diff
@@ -0,0 +1,492 @@
+// promise-style. Most of this will be removed once the promises conversion is
+// completed.
+
+#include <grpc/support/port_platform.h>
+
+#include "absl/utility/utility.h"
+
+#include <grpc/status.h>
+#include <grpc/support/log.h>
+
+#include "src/core/lib/channel/channel_stack.h"
+#include "src/core/lib/promise/arena_promise.h"
+#include "src/core/lib/promise/promise.h"
+#include "src/core/lib/transport/error_utils.h"
+
+namespace grpc_core {
+
+namespace promise_filter_detail {
+class BaseCallData;
+};
+
+// Small unowned "handle" type to ensure one accessor at a time to metadata.
+// The focus here is to get promises to use the syntax we'd like - we'll
+// probably substitute some other smart pointer later.
+template <typename T>
+class MetadataHandle {
+ public:
+  MetadataHandle() = default;
+
+  MetadataHandle(const MetadataHandle&) = delete;
+  MetadataHandle& operator=(const MetadataHandle&) = delete;
+
+  MetadataHandle(MetadataHandle&& other) noexcept : handle_(other.handle_) {
+    other.handle_ = nullptr;
+  }
+  MetadataHandle& operator=(MetadataHandle&& other) noexcept {
+    handle_ = other.handle_;
+    other.handle_ = nullptr;
+    return *this;
+  }
+
+  T* operator->() const { return handle_; }
+  bool has_value() const { return handle_ != nullptr; }
+
+  static MetadataHandle TestOnlyWrap(T* p) { return MetadataHandle(p); }
+
+ private:
+  friend class promise_filter_detail::BaseCallData;
+
+  explicit MetadataHandle(T* handle) : handle_(handle) {}
+  T* Unwrap() {
+    T* result = handle_;
+    handle_ = nullptr;
+    return result;
+  }
+
+  T* handle_ = nullptr;
+};
+
+// Trailing metadata type
+// TODO(ctiller): This should be a bespoke instance of MetadataMap<>
+using TrailingMetadata = MetadataHandle<grpc_metadata_batch>;
+
+// Initial metadata type
+// TODO(ctiller): This should be a bespoke instance of MetadataMap<>
+using InitialMetadata = MetadataHandle<grpc_metadata_batch>;
+
+using NextPromiseFactory =
+    std::function<ArenaPromise<TrailingMetadata>(InitialMetadata)>;
+
+namespace promise_filter_detail {
+
+// Call data shared between all implementations of promise-based filters.
+class BaseCallData {
+ public:
+  BaseCallData(grpc_call_element* elem, const grpc_call_element_args* args)
+      : elem_(elem),
+        arena_(args->arena),
+        call_combiner_(args->call_combiner),
+        owning_call_(args->call_stack),
+        deadline_(args->deadline) {}
+
+ protected:
+  class ScopedContext : public promise_detail::Context<Arena> {
+   public:
+    explicit ScopedContext(BaseCallData* call_data)
+        : promise_detail::Context<Arena>(call_data->arena_) {}
+  };
+
+  static MetadataHandle<grpc_metadata_batch> WrapMetadata(
+      grpc_metadata_batch* p) {
+    return MetadataHandle<grpc_metadata_batch>(p);
+  }
+
+  static grpc_metadata_batch* UnwrapMetadata(
+      MetadataHandle<grpc_metadata_batch> p) {
+    return p.Unwrap();
+  }
+
+  grpc_call_element* const elem_;
```

Comment:
Not all filters actually store all of these values.  By storing these once for every promise-based filter, are we going to increase memory usage in the short term?  I realize that in the long term, these will be stored once across all filters instead of once per filter, so it will actually decrease memory usage, but I'm wondering if the short-term increase is going to be a problem.

Could we ameliorate this by providing a call context struct that we pass down through the filter stack, and then use that here?  In principle, the change to have a call context object seems orthogonal to the switch from filters to promises.

Reply:
There are multiple paths to removing this overhead, and which we choose is going to depend a little on how the next steps go (I don't have enough information to choose):
- we can, as you say, introduce some context that we set somewhere and use it
- we can share this overhead between several filters that are inline
- we can do something else that becomes apparent when we have more code converted
- we can wrap a promise around a filter once most of the stack is converted (the inverse of this code)

There'll definitely be a short term increase to watch out for - and we'll likely deploy some mitigations at some point - but I want to wait to decide what.

Reply:
Okay, I'm fine with deferring this for now.  Just wanted to make sure we were thinking about it.